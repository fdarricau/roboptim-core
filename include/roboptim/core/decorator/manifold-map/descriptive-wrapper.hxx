#ifndef ROBOPTIM_CORE_DECORATOR_MANIFOLD_MAP_DESCRIPTIVE_WRAPPER_HXX
# define ROBOPTIM_CORE_DECORATOR_MANIFOLD_MAP_DESCRIPTIVE_WRAPPER_HXX
# include <boost/format.hpp>
# include <manifolds/Manifold.h>
# include "manifold-desc.hh"

namespace roboptim
{

  template <typename U, typename V>
  template<class ... Types>
  DescriptiveWrapper<U, V>::DescriptiveWrapper
  (Types ... args)
  {
    U* stdPtr = new U(args...);
    this->fct_ = boost::shared_ptr<U>(stdPtr);
    this->manifold_ = boost::shared_ptr<pgs::Manifold>(V::getManifold( &(*(this->fct_))));
  }

  template <typename U, typename V>
  DescriptiveWrapper<U, V>::DescriptiveWrapper
  (boost::shared_ptr<U>& fct, const pgs::Manifold& manifold)
  : fct_ (fct),
    manifold_ (&manifold)
  {
    long size = manifold_->representationDim();
    if (fct_->inputSize() != size)
    {
      std::stringstream* error = new std::stringstream;
      (*error) << "Representation dims mismatch on manifold "
               << manifold_->name() << " using function "
               << fct_->getName() << ". Expected dimension :" << size
               << ", actual one is " << fct_->inputSize();
      throw std::runtime_error (error->str());
    }
  }

  template <typename U, typename V>
  DescriptiveWrapper<U, V>::~DescriptiveWrapper()
  {
  }
} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_DECORATOR_MANIFOLD_MAP_DESCRIPTIVE_WRAPPER_HXX
