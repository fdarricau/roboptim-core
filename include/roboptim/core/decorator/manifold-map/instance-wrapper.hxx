#ifndef ROBOPTIM_CORE_DECORATOR_MANIFOLD_MAP_INSTANCE_WRAPPER_HXX
# define ROBOPTIM_CORE_DECORATOR_MANIFOLD_MAP_INSTANCE_WRAPPER_HXX
# include <vector>
# include <queue>
# include <utility>
# include <iostream>
# include <typeinfo>

# include <boost/format.hpp>
# include <boost/mpl/assert.hpp>

# include <roboptim/core/decorator/manifold-map/descriptive-wrapper.hh>

# include <manifolds/Manifold.h>
# include <manifolds/RealSpace.h>
# include <manifolds/S2.h>
namespace roboptim
{

  template <typename U>
  InstanceWrapper<U>::~InstanceWrapper()

  {
    delete [] this->mappingFromFunction_;
  }

  template <typename U>
  void
  InstanceWrapper<U>::mapArgument (const_argument_ref argument)
    const
  {
    for (long i = 0; i < this->mappingFromFunctionSize_; ++i)
      {
	this->mappedInput_(i) = argument(static_cast<long>(this->mappingFromFunction_[i]));
      }
  }

  template <typename U>
  void
  InstanceWrapper<U>::unmapGradient(gradient_ref gradient)
    const
  {
    for (long i = 0; i < this->mappingFromFunctionSize_; ++i)
      {
	gradient(static_cast<long>(this->mappingFromFunction_[i])) = this->mappedGradient_(i);
      }
  }

  template <typename U>
  void
  InstanceWrapper<U>::impl_compute
  (result_ref result, const_argument_ref x)
    const
  {
    this->mapArgument(x);
    (*this->fct_)(result, this->mappedInput_);
  }

  template <typename U>
  void
  InstanceWrapper<U>::impl_gradient (gradient_ref gradient,
			 const_argument_ref argument,
			 size_type functionId)
    const
  {
    this->mapArgument(argument);
    this->fct_->gradient(this->mappedGradient_, this->mappedInput_, functionId);

    gradient.setZero();
    this->unmapGradient(gradient);
  }

  template <typename U>
  void
  InstanceWrapper<U>::impl_jacobian (jacobian_ref jacobian,
			 const_argument_ref argument)
    const
  {
    this->mapArgument(argument);
    jacobian.setZero();

    for (long j = 0; j < jacobian.rows(); ++j)
      {
	this->fct_->gradient(this->mappedGradient_, this->mappedInput_, j);
	this->unmapGradient(jacobian.row(j));
      }
  }

  template <typename U>
  std::ostream&
  InstanceWrapper<U>::print_ (std::ostream& o)
  {
    for (long i = 0; i < this->mappingFromFunctionSize_; ++i)
      {
	o << (i>0?", ":"") << this->mappingFromFunction_[i];
      }

    return o;
  }


} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_DECORATOR_MANIFOLD_MAP_INSTANCE_WRAPPER_HXX
