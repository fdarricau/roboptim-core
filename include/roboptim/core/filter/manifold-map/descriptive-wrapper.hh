#ifndef ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_DESCRIPTIVE_WRAPPER_HH
# define ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_DESCRIPTIVE_WRAPPER_HH
# include <vector>
# include <ostream>
# include <boost/shared_ptr.hpp>

# include <roboptim/core/differentiable-function.hh>

# include <manifolds/Manifold.h>

namespace roboptim
{
  /// \addtogroup roboptim_filter
  /// @{

  // FIXME: private constructor MAGIC

  /// \brief apply a given roboptim function to a given dimension space
  /// \tparam U input function type.
  template <typename U, typename V>
  class DescriptiveWrapper
  {
  public:
    typedef boost::shared_ptr<DescriptiveWrapper> DescriptiveWrapperShPtr_t;

    /// \brief map the function on the given manifold.
    /// \param fct input function.
    /// \param functionManifold the manifold describing the function's input vector.
    template<class ... Types>
    explicit DescriptiveWrapper (Types ... args);

    DescriptiveWrapper (boost::shared_ptr<U>& f, const pgs::Manifold& m);

    ~DescriptiveWrapper ();

    const pgs::Manifold& manifold () const
    {
      return *manifold_;
    }

    const boost::shared_ptr<U>& fct () const
    {
      return fct_;
    }

    U& fct ()
    {
      return *fct_;
    }

  private:

    boost::shared_ptr<U>  fct_;
    const pgs::Manifold* manifold_;
  };

  template <typename U, typename V>
  boost::shared_ptr<DescriptiveWrapper<U, V> >
  descriptivewrapper (boost::shared_ptr<U> fct,
		      const pgs::Manifold& manifold)
  {
    return boost::make_shared<DescriptiveWrapper<U, V> > (fct, manifold);
  }

  template <typename U, typename V>
  std::ostream&
  operator<<(std::ostream& o, DescriptiveWrapper<U, V>& descWrap)
  {
    o <<
    "Displaying info about function in DescriptiveWrapper :" << "\n" <<
    "the function is " << descWrap.fct().getName() << "\n" <<
    "the function input size is " << descWrap.fct().inputSize() << "\n" <<
    "the function output size is " << descWrap.fct().outputSize() << "\n" <<
    "Displaying info about the function manifold in DescriptiveWrapper :" << "\n" <<
    "the manifold is " << descWrap.manifold().name() << "\n" <<
    "Is the manifold elementary ? : " <<
    (descWrap.manifold().isElementary() ? "yes" : "no" ) << "\n" <<
    "the manifold dimension is " << descWrap.manifold().representationDim() << "\n";
    return o;
  }

} // end of namespace roboptim.

# include <roboptim/core/filter/manifold-map/descriptive-wrapper.hxx>
#endif //! ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_DESCRIPTIVE_WRAPPER_HH
