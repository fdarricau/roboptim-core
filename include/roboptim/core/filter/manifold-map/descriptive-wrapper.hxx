#ifndef ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_DESCRIPTIVE_WRAPPER_HXX
# define ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_DESCRIPTIVE_WRAPPER_HXX
# include <boost/format.hpp>
# include <manifolds/Manifold.h>

namespace roboptim
{

  template <typename U, typename V>
  template<class ... Types>
  DescriptiveWrapper<U, V>::DescriptiveWrapper
  (Types ... args)
  {
    // Creation

    this->fct_ = new U(args...);
    V manifoldGenerator;
    this->manifold_ = manifoldGenerator.getManifold(this->function_);

    // Manifold checking should be useless now, commenting
    /*
    long size = manifold_->representationDim();
    if (fct_->inputSize() != size)
    {
      std::stringstream* error = new std::stringstream;
      (*error) << "Representation dims mismatch on manifold "
               << manifold_->name() << " using function "
               << fct->getName() << ". Expected dimension :" << size
               << ", actual one is " << fct_->inputSize();
      throw std::runtime_error (error->str());
    }*/
  }

  template <typename U>
  DescriptiveWrapper<U>::~DescriptiveWrapper()
  {
  }
} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_DESCRIPTIVE_WRAPPER_HXX
