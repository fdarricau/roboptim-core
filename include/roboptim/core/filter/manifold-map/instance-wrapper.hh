#ifndef ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_INSTANCE_WRAPPER_HH
# define ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_INSTANCE_WRAPPER_HH
# include <vector>
# include <iostream>
# include <boost/shared_ptr.hpp>

# include <roboptim/core/detail/autopromote.hh>
# include <roboptim/core/differentiable-function.hh>

# include <manifolds/Manifold.h>


namespace roboptim
{
  /// \addtogroup roboptim_filter
  /// @{

  /// \brief Binds a DescriptiveWrapper to a instance of a submanifold.
  /// \tparam U input function type.
  template <typename U>
  class InstanceWrapper : public detail::AutopromoteTrait<U>::T_type
  {
  public:
    typedef typename detail::AutopromoteTrait<U>::T_type parentType_t;
    ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_ (parentType_t);

    typedef boost::shared_ptr<InstanceWrapper> InstanceWrapperShPtr_t;

    /// \brief Create a mapping from the problem's manifold to the function's.
    /// \param fct input function.
    /// \param problemManifold the manifold describing the whole variable vector.
    /// \param functionManifold the manifold describing the function's input vector.
    explicit InstanceWrapper (boost::shared_ptr<U> fct,
			  pgs::Manifold& problemManifold,
			  pgs::Manifold& functionManifold);
    ~InstanceWrapper ();

    const boost::shared_ptr<U>& origin () const
    {
      return origin_;
    }

    boost::shared_ptr<U>& origin ()
    {
      return origin_;
    }

    void impl_compute (result_ref result, const_argument_ref x)
      const;

    void impl_gradient (gradient_ref gradient,
			const_argument_ref argument,
			size_type functionId = 0)
      const;
    void impl_jacobian (jacobian_ref jacobian,
			const_argument_ref arg)
      const;

    std::ostream& print(std::ostream& o);
  private:
    boost::shared_ptr<U> origin_;

    int* mappingFromProblem_;
    size_t mappingFromProblemSize_;
    int* mappingFromFunction_;
    size_t mappingFromFunctionSize_;
  };

  template <typename U>
  boost::shared_ptr<InstanceWrapper<U> >
  scalar (boost::shared_ptr<U> origin,
	  typename InstanceWrapper<U>::size_type start = 0,
	  typename InstanceWrapper<U>::size_type size = 1)
  {
    return boost::make_shared<InstanceWrapper<U> > (origin, start, size);
  }

  template <typename U>
  boost::shared_ptr<InstanceWrapper<U> >
  operator* (typename InstanceWrapper<U>::value_type scalar,
	     boost::shared_ptr<U> origin)
  {
    return boost::make_shared<InstanceWrapper<U> > (origin, scalar);
  }

  template <typename U>
  boost::shared_ptr<InstanceWrapper<U> >
  operator* (boost::shared_ptr<U> origin,
	     typename InstanceWrapper<U>::value_type scalar)
  {
    return boost::make_shared<InstanceWrapper<U> > (origin, scalar);
  }

  template <typename U>
  boost::shared_ptr<U>
  operator+ (boost::shared_ptr<U> origin)
  {
    return origin;
  }

  template <typename U>
  boost::shared_ptr<InstanceWrapper<U> >
  operator- (boost::shared_ptr<U> origin)
  {
    return boost::make_shared<InstanceWrapper<U> > (origin, -1.);
  }

  template <typename U>
  std::ostream&
  operator<<(std::ostream& o, InstanceWrapper<U>& instWrap)
  {
    return instWrap.print(o);
  }

  /// @}

} // end of namespace roboptim.


# include <roboptim/core/filter/manifold-map/instance-wrapper.hxx>
#endif //! ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_INSTANCE_WRAPPER_HH
