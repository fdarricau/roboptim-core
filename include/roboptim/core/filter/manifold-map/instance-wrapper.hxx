#ifndef ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_INSTANCE_WRAPPER_HXX
# define ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_INSTANCE_WRAPPER_HXX
# include <boost/format.hpp>
# include <vector>
# include <queue>
# include <manifolds/Manifold.h>
# include <iostream>

# include <roboptim/core/filter/manifold-map/descriptive-wrapper.hh>

namespace roboptim
{

  template <typename U>
  InstanceWrapper<U>::InstanceWrapper
  (boost::shared_ptr<DescriptiveWrapper<U>> descWrap,
     pgs::Manifold& problemManifold,
     pgs::Manifold& functionManifold)
    : detail::AutopromoteTrait<U>::T_type
      (descWrap->fct().inputSize (),
       descWrap->fct().outputSize (),
       (boost::format ("%1%")
	% descWrap->fct().getName ()).str ()),
      descWrap_ (descWrap)
  {
    this->mappingFromProblemSize_ = problemManifold.representationDim();
    this->mappingFromProblem_ = new size_t[this->mappingFromProblemSize_];
    this->mappingFromFunctionSize_ = functionManifold.representationDim();
    this->mappingFromFunction_ = new size_t[this->mappingFromFunctionSize_];

    std::cout << "this->mappingFromProblemSize_: " << this->mappingFromProblemSize_ << std::endl;
    std::cout << "this->mappingFromFunctionSize_: " << this->mappingFromFunctionSize_ << std::endl;

    std::function<long(const pgs::Manifold&, std::string, long, long)> getStartingIndexOfManifold = [&getStartingIndexOfManifold, this](const pgs::Manifold& manifold, std::string targetName, long functionStartIndex, long startIndex)
      {
	if (!targetName.compare(manifold.name()))
	  {
	    // We found the manifold
	    // Write its index and exit
	    std::cout << "Found a match for " << manifold.name()  << std::endl;
	    std::cout << "startIndex: " << startIndex << std::endl;
	    std::cout << "functionStartIndex: " << functionStartIndex << std::endl;

	    for (long i = 0; i < manifold.representationDim(); ++i)
	      {
		std::cout << "i: " << i << std::endl;
		this->mappingFromProblem_[static_cast<size_t>(startIndex + i)] = static_cast<size_t>(functionStartIndex + i);
		this->mappingFromFunction_[static_cast<size_t>(functionStartIndex + i)] = static_cast<size_t>(startIndex + i);
	      }

	    return -1l;
	  }
	else
	  {
	    if (manifold.isElementary())
	      {
		return startIndex + manifold.representationDim();
	      }
	    else
	      {
		for (size_t i = 0; i < manifold.numberOfSubmanifolds() && startIndex >= 0; ++i)
		  {
		    startIndex = getStartingIndexOfManifold(manifold(i), targetName, functionStartIndex, startIndex);
		  }

		return startIndex;
	      }
	  }
      };

    std::function<int(const pgs::Manifold&, int)> traverseFunctionManifold = [&traverseFunctionManifold, &getStartingIndexOfManifold, &problemManifold](const pgs::Manifold& manifold, int startIndex)
      {
	if (manifold.isElementary())
	  {
	    std::cout << "Matching function manifold: " << manifold.name()  << std::endl;
	    getStartingIndexOfManifold(problemManifold, manifold.name(), startIndex, 0);
	    return static_cast<int>(startIndex + manifold.representationDim());
	  }

	for (size_t i = 0; i < manifold.numberOfSubmanifolds() && startIndex >= 0; ++i)
	  {
	    startIndex = traverseFunctionManifold(manifold(i), startIndex);
	  }

	return static_cast<int>(startIndex);
      };

    traverseFunctionManifold(functionManifold, 0);
  }

  template <typename U>
  InstanceWrapper<U>::~InstanceWrapper()
  {
    delete [] this->mappingFromProblem_;
    delete [] this->mappingFromFunction_;
  }

  // FIXME: temp, complete those methods
  template <typename U>
  void
  InstanceWrapper<U>::mapArgument (const_argument_ref argument)
    const
  {
    for (long i = 0; i < this->mappingFromFunctionSize_; ++i)
      {
	this->mappedInput_(i) = argument(static_cast<long>(this->mappingFromFunction_[i]));
      }
  }

  template <typename U>
  void
  InstanceWrapper<U>::unmapGradient(gradient_t gradient)
    const
  {
    for (long i = 0; i < this->mappingFromFunctionSize_; ++i)
      {
	gradient(static_cast<long>(this->mappingFromFunction_[i])) = this->mappedGradient_(i);
      }
  }

  template <typename U>
  void
  InstanceWrapper<U>::impl_compute
  (result_ref result, const_argument_ref x)
    const
  {
    this->mapArgument(x);

    descWrap_->fct()(result, this->mappedInput_);
  }

  template <typename U>
  void
  InstanceWrapper<U>::impl_gradient (gradient_ref gradient,
			 const_argument_ref argument,
			 size_type functionId)
    const
  {
    this->mapArgument(argument);
    descWrap_->fct().gradient(this->mappedGradient_, this->mappedInput_, functionId);

    gradient.setZero();
    this->unmapGradient(gradient);
  }

  template <typename U>
  void
  InstanceWrapper<U>::impl_jacobian (jacobian_ref jacobian,
			 const_argument_ref argument)
    const
  {
    this->mapArgument(argument);
    jacobian.setZero();

    for (long j = 0; j < jacobian.cols(); ++j)
      {
	descWrap_->fct().gradient(this->mappedGradient_, this->mappedInput_, j);
	this->unmapGradient(jacobian.col(j));
      }

  }

  template <typename U>
  std::ostream&
  InstanceWrapper<U>::print_ (std::ostream& o)
  {
    for (long i = 0; i < this->mappingFromProblemSize_; ++i)
      {
	o << (i>0?", ":"") << this->mappingFromProblem_[i];
      }
    o << "\n";
    for (long i = 0; i < this->mappingFromFunctionSize_; ++i)
      {
	o << (i>0?", ":"") << this->mappingFromFunction_[i];
      }
    o << "\n";

    return o;
  }


} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_FILTER_MANIFOLD_MAP_INSTANCE_WRAPPER_HXX
