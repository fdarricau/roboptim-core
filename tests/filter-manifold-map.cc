// Copyright (C) 2013 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>

#include <roboptim/core/differentiable-function.hh>

#include <roboptim/core/filter/manifold-map/instance-wrapper.hh>
#include <roboptim/core/filter/manifold-map/descriptive-wrapper.hh>

#include <manifolds/SO3.h>
#include <manifolds/RealSpace.h>
#include <manifolds/CartesianProduct.h>
#include <manifolds/ExpMapMatrix.h>
#include <manifolds/S2.h>

using namespace roboptim;

typedef boost::mpl::list< ::roboptim::EigenMatrixDense,
  ::roboptim::EigenMatrixSparse> functionTypes_t;

struct F : public DifferentiableFunction
{
  F () : DifferentiableFunction (22, 10, "f_n (x) = n * x")
  {}

  void impl_compute (result_ref res, const_argument_ref argument) const
  {
    res.setZero ();
    for (size_type i = 0; i < outputSize (); ++i)
      for (size_type j = 0; j < 3; ++j)
	{
	  res[i] += (value_type)i * argument[19 + j];
	}
  }

  void impl_gradient (gradient_ref grad, const_argument_ref,
		      size_type functionId) const
  {
    grad.setZero ();
    for (size_type j = 0; j < 3; ++j)
      {
	grad[19 + j] += (value_type)functionId;
      }
  }
};

struct G : public DifferentiableFunction
{
  G () : DifferentiableFunction (3, 1, "f_n (x) = sum(x)")
  {}

  void impl_compute (result_ref res, const_argument_ref argument) const
  {
    res.setZero ();
    for (size_type i = 0; i < inputSize (); ++i)
      {
	res[0] += argument[i];
      }
  }

  void impl_gradient (gradient_ref grad, const_argument_ref,
		      size_type functionId) const
  {
    grad.setZero ();
    for (size_type j = 0; j < 3; ++j)
      grad[j] = functionId * 0 + 1;
  }
};

struct H : public DifferentiableFunction
{
  H () : DifferentiableFunction (45, 3, "f_n (x) = sum(x)")
  {}

  void impl_compute (result_ref res, const_argument_ref argument) const
  {
    res.setZero ();
    for (size_type i = 0; i < inputSize (); ++i)
      {
	res[i % 3] += argument[i];
      }
  }

  void impl_gradient (gradient_ref grad, const_argument_ref,
		      size_type functionId) const
  {
    grad.setZero ();
    for (size_type j = 0; j < 45; ++j)
      grad[j] = (functionId == (j % 3));
  }
};

boost::shared_ptr<boost::test_tools::output_test_stream> output;

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE (manifold_map_test_0)
{/*
  output = retrievePattern("filter-manifold-map");

  boost::shared_ptr<F> f (new F());

  pgs::RealSpace pos(3);pos.name() = "position";
  pgs::SO3<pgs::ExpMapMatrix> ori; ori.name() = "orientation";
  const pgs::CartesianProduct freeFlyer(pos, ori);
  pgs::RealSpace joints(10); joints.name() = "joints";
  const pgs::CartesianProduct robot(freeFlyer, joints);

  const pgs::CartesianProduct cartProd(joints, ori);
  const pgs::CartesianProduct myFuncManifold(cartProd, pos);

  boost::shared_ptr<DescriptiveWrapper<DifferentiableFunction>>
    descWrapPtr(new DescriptiveWrapper<DifferentiableFunction>(f, myFuncManifold));

  InstanceWrapper<DifferentiableFunction> instWrap(descWrapPtr, robot, myFuncManifold);

  InstanceWrapper<DifferentiableFunction>::argument_t input = Eigen::VectorXd::Zero(22);
  InstanceWrapper<DifferentiableFunction>::result_t result = Eigen::VectorXd::Zero(10);
  InstanceWrapper<DifferentiableFunction>::gradient_t gradient = Eigen::VectorXd::Zero(22);
  InstanceWrapper<DifferentiableFunction>::jacobian_t jacobian = Eigen::MatrixXd::Zero(10, 22);

  for(int i = 0; i < 3; ++i)
    {
      input(i) = 1 + i;
    }

  (*output) << instWrap << "\n";
  std::cout << instWrap << std::endl;
  std::cout << "input: " << input.transpose() << std::endl;
  instWrap(result, input);
  std::cout << "result: " << result.transpose() << std::endl << std::endl;

  for (int i = 0; i < f->outputSize(); ++i)
    {
      instWrap.gradient(gradient, input, i);
    }

  instWrap.jacobian(jacobian, input);
  std::cout << "jacobian: " << std::endl << jacobian << std::endl;

  (*output) << (*descWrapPtr);

  BOOST_CHECK (output->match_pattern());

}

BOOST_AUTO_TEST_CASE (manifold_map_test_1)
{
  output = retrievePattern("filter-manifold-map-1");

  boost::shared_ptr<G> g (new G());

  std::vector<const pgs::RealSpace*> reals;
  pgs::CartesianProduct problemManifold;
  const pgs::RealSpace descriptiveManifold(3);

  boost::shared_ptr<DescriptiveWrapper<DifferentiableFunction>>
    descWrapPtr(new DescriptiveWrapper<DifferentiableFunction>(g, descriptiveManifold));

  size_t posNumber = 15;

  for (size_t i = 0; i < posNumber; ++i)
    {
      pgs::RealSpace* newR = new pgs::RealSpace(3);
      newR->name() = "position (" + std::to_string(i) + ")";
      reals.push_back(newR);
      problemManifold.multiply(*reals.back());
    }

  InstanceWrapper<DifferentiableFunction>::argument_t input = Eigen::VectorXd::Zero(3 * static_cast<long>(posNumber));
  InstanceWrapper<DifferentiableFunction>::result_t result = Eigen::VectorXd::Zero(1);
  InstanceWrapper<DifferentiableFunction>::gradient_t gradient = Eigen::VectorXd::Zero(3 * static_cast<long>(posNumber));
  InstanceWrapper<DifferentiableFunction>::jacobian_t jacobian = Eigen::MatrixXd::Zero(1, 3 * static_cast<long>(posNumber));

  for (int i = 0; i < input.size(); ++i)
    {
      input(i) = i;
    }

  for (size_t i = 0; i < posNumber; ++i)
    {
      InstanceWrapper<DifferentiableFunction> instWrap(descWrapPtr, problemManifold, *reals[i]);

      instWrap(result, input);

      BOOST_CHECK(result(0) == (3 * (3 * i + 1)));

      instWrap.jacobian(jacobian, input);
      std::cout << jacobian << std::endl;
      (*output) << jacobian << std::endl;
    }

  for (size_t i = 0; i < posNumber; ++i)
    {
      delete reals[i];
    }

  }

BOOST_AUTO_TEST_CASE (manifold_map_test_2)
{
  output = retrievePattern("filter-manifold-map-2");

  boost::shared_ptr<F> f (new F());
  boost::shared_ptr<G> g (new G());

  const pgs::RealSpace manifold(2);

  try
  {
    boost::shared_ptr<DescriptiveWrapper<DifferentiableFunction>>
    descWrapPtr(new DescriptiveWrapper<DifferentiableFunction>(g, manifold));
  }
  catch (std::runtime_error& e)
  {
    (*output) << "std::runtime_error: " << e.what() << "\n";
  }

  const pgs::SO3<pgs::ExpMapMatrix> manifold2;
  try
  {
    boost::shared_ptr<DescriptiveWrapper<DifferentiableFunction>>
    descWrapPtr(new DescriptiveWrapper<DifferentiableFunction>(f, manifold2));
  }
  catch (std::runtime_error& e)
  {
    (*output) << "std::runtime_error: " << e.what() << "\n";
  }

  const pgs::CartesianProduct manifold3(manifold, manifold2);

  try
  {
    boost::shared_ptr<DescriptiveWrapper<DifferentiableFunction>>
    descWrapPtr(new DescriptiveWrapper<DifferentiableFunction>(g, manifold3));
  }
  catch (std::runtime_error& e)
  {
    (*output) << "std::runtime_error: " << e.what() << "\n";
  }

  BOOST_CHECK (output->match_pattern());

}

BOOST_AUTO_TEST_CASE (manifold_map_test_3)
{
  output = retrievePattern("filter-manifold-map-3");

  boost::shared_ptr<H> h (new H());

  std::vector<const pgs::Manifold*> reals;
  std::vector<std::pair<long, long>> restrictions;
  pgs::CartesianProduct problemManifold;
  pgs::CartesianProduct descriptiveManifold;

  const size_t posNumber = 15;

  for (size_t i = 0; i < posNumber; ++i)
    {
      pgs::RealSpace* newR = new pgs::RealSpace(6);
      newR->name() = "position (" + std::to_string(i) + ")";
      reals.push_back(newR);
      problemManifold.multiply(*reals.back());
      descriptiveManifold.multiply(*(new pgs::RealSpace(3)));
    }

  restrictions.push_back(std::make_pair(3l, 3l));

  boost::shared_ptr<DescriptiveWrapper<DifferentiableFunction>>
    descWrapPtr(new DescriptiveWrapper<DifferentiableFunction>(h, descriptiveManifold));

  InstanceWrapper<DifferentiableFunction>::argument_t input = Eigen::VectorXd::Zero(6 * static_cast<long>(posNumber));
  InstanceWrapper<DifferentiableFunction>::result_t result = Eigen::VectorXd::Zero(3);
  InstanceWrapper<DifferentiableFunction>::gradient_t gradient = Eigen::VectorXd::Zero(6 * static_cast<long>(posNumber));
  InstanceWrapper<DifferentiableFunction>::jacobian_t jacobian = Eigen::MatrixXd::Zero(3, 6 * static_cast<long>(posNumber));

  InstanceWrapper<DifferentiableFunction> instWrap(descWrapPtr, problemManifold, problemManifold, reals, restrictions);

 for (int i = 0; i < input.size(); ++i)
    {
      input(i) = (i % 6) > 2;
    }

 instWrap(result, input);

 (*output) << instWrap;

 std::cout << "result: " << result << std::endl;

 BOOST_CHECK (result(0) == 15);
 BOOST_CHECK (result(1) == 15);
 BOOST_CHECK (result(2) == 15);

 instWrap.jacobian(jacobian, input);

 Eigen::MatrixXd jac = jacobian;

 (*output) << "\n";

 for (int i = 0; i < jac.rows(); ++i)
   {
     std::cout << jac.row(i) << std::endl;
     (*output) << jac.row(i) << std::endl;
   }

 BOOST_CHECK (output->match_pattern());
}

BOOST_AUTO_TEST_CASE (manifold_map_test_4)
{
  boost::shared_ptr<F> f (new F());

  pgs::RealSpace pos(3);pos.name() = "position";
  pgs::SO3<pgs::ExpMapMatrix> ori; ori.name() = "orientation";
  const pgs::CartesianProduct freeFlyer(pos, ori);
  pgs::RealSpace joints(10); joints.name() = "joints";
  const pgs::CartesianProduct robot(freeFlyer, joints);

  const pgs::S2 s2;
  const pgs::CartesianProduct cartProd(joints, ori);
  const pgs::CartesianProduct myFuncManifold(cartProd, s2);
  const pgs::CartesianProduct mySubManifold(cartProd, pos);

  boost::shared_ptr<DescriptiveWrapper<DifferentiableFunction>>
    descWrapPtr(new DescriptiveWrapper<DifferentiableFunction>(f, myFuncManifold));

  std::vector<const pgs::Manifold*> restrictedManifolds;
  restrictedManifolds.push_back(&pos);
  std::vector<std::pair<long, long>> restrictions;
  restrictions.push_back(std::make_pair(1l, 1l));

  bool errorThrown = false;

  try
    {
      InstanceWrapper<DifferentiableFunction> instWrap(descWrapPtr, robot, mySubManifold);
    }
  catch (std::runtime_error& e)
    {
      errorThrown = true;
    }

  BOOST_CHECK(errorThrown);
  errorThrown = false;

    try
    {
      InstanceWrapper<DifferentiableFunction> instWrap(descWrapPtr, robot, mySubManifold, restrictedManifolds, restrictions);
    }
  catch (std::runtime_error& e)
    {
      errorThrown = true;
    }

  BOOST_CHECK(errorThrown);
*/}


BOOST_AUTO_TEST_SUITE_END ()
