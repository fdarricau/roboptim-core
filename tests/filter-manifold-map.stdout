12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 3, 4, 5, 6, 7, 8, 9, 10, 11, 0, 1, 2
Displaying info about function in DescriptiveWrapper :
the function is f_n (x) = n * x
the function input size is 22
the function output size is 10
Displaying info about the function manifold in DescriptiveWrapper :
the manifold is R10xSO3xR3
Is the manifold elementary ? : no
the manifold dimension is 22
